### PR TITLE
feat: add adapter for feishu (飞书)

### PR DESCRIPTION
**Pull request by MaikoTan**
Superseded from https://github.com/koishijs/koishi/pull/663

Feishu (飞书) provide native chat bot support, but it is included in "Tenant Application" (企业自建应用) or "Market Applications" (应用商店应用).

Donc, it is required to create a "Tenant Application" before enabling the "Robot Ability" (机器人能力) and subscribe to (at least) "Message Receive" events.
See also: https://open.feishu.cn/document/home/develop-a-bot-in-5-minutes/create-an-app (Chinese)

------

# Features

### API

- [x] Receive and decrypt messages
- [x] Send and encrypt messages
- [x] Recall messages
- [x] Get read users of messages

#### Planned

- [x] Text (both send and receive, as following)
- [x] Image (upload, send, receive, download)
- [x] Audio
- [x] File
- [x] Media (Video)
- [x] Post (Rich Text)

#### Currently Not Planned (Subject to change)

- Sticker (Emoji) (Don't know why but not a notable segment in messages so not able to supported)
- Interactive (Interactive Card, which allow user to click buttons or something)
- Hongbao (红包 or Otoshidama)
- Share Calendar Event (Sharable Schedule)
- Share Chat (Sharable Contact Card for Group)
- Share User (Sharable Contact Card for User)
- System
- Location
- Video Chat

## Other APIs

- [x] Get Tokens
- [x] Get groups
- [x] Get users in groups

### CODE REVIEW

File: adapters/feishu/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import * as Feishu from './types'
+import { FeishuBot } from './bot'
+
+export { Feishu }
+
+export default FeishuBot
+
+declare module '@satorijs/core' {
+  interface Session {
+    Feishu: Feishu.Internal
```
Comment (Anillc) [inline adapters/feishu/src/index.ts:10]: 这里是不是应该改为小写的 `feishu`, 以及 `adaptSession` 中好像并没有给这个属性赋值，是否也应该把 `Feishu.Event` 也加进去

```suggestion
    feishu: Feishu.Internal & Feishu.Event
```

File: adapters/feishu/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import crypto from 'crypto'
+
+import { Context, Message, segment, Session } from '@satorijs/satori'
+
+import { FeishuBot } from './bot'
+import { Event, MessageContentType } from './types'
+
+export function adaptMessage(bot: FeishuBot, data: Event<'im.message.receive_v1'>['event']): Message {
+  const json = JSON.parse(data.message.content) as MessageContentType<typeof data.message.message_type>
+  const assetEndpoint = bot.config.selfUrl + bot.config.path + '/assets'
```
Comment (Anillc) [inline adapters/feishu/src/utils.ts:10]: 这里的 `selfUrl` 也有可能没有配置，是不是应该在启动的时候检查一下。并且除了插件配置中的 `selfUrl`，全局配置中也有一个 `selfUrl`，应该在插件的 `selfUrl` 没有配置的时候使用全局的 `selfUrl`

Comment (MaikoTan) [reply adapters/feishu/src/utils.ts:10]: 確實，但這裡我並不是很確定默認的 selfUrl 會從哪裡讀入，它會是作為 `bot.selfUrl` 嗎？

File: adapters/feishu/src/bot.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            logger.warn(`@here or @role{${data.role}} is not supported`)
+          }
+          break
+        }
+        case 'image':
+        case 'audio':
+        case 'video':
+        case 'file': {
+          await sendBuffer()
+          const content = await this._prepareAssets(type, data)
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply adapters/feishu/src/bot.ts:150]: 此處可能需要修改整個 Event 的類型，我原本期待的是能夠在前面的 if-check 之後立刻 type narrowing，但很明顯沒有。也許泛型應該寫在 EventBody 上？

File: adapters/feishu/src/index.ts (hunk 1 above)
Comment (MaikoTan) [reply adapters/feishu/src/index.ts:10]: 是的，這裡應該是小寫。

這裡加入 Feishu.Event 的目的是什麼？

File: adapters/feishu/src/bot.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  set token(v: string) {
+    this._token = v
+    this.http.config.headers.Authorization = `Bearer ${v}`
+  }
+
+  async sendMessage(channelId: string, content: string, guildId?: string): Promise<string[]> {
+    const session = await this.session({ channelId, content, guildId, subtype: guildId ? 'group' : 'private' })
+    if (!session?.content) return []
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply adapters/feishu/src/bot.ts]: Done

File: adapters/feishu/src/http.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    logger.debug('http server stopped')
+  }
+
+  async dispatchSession(body: Event): Promise<void> {
+    const { header } = body
+    const { app_id } = header
+    const bot = this.bots.find((bot) => bot.selfId === app_id)
+    const session = await adaptSession(bot, body)
+    bot.dispatch(session)
+  }
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply adapters/feishu/src/http.ts]: Done

File: adapters/feishu/src/utils.ts (hunk 2 above)
Comment (Anillc) [reply adapters/feishu/src/utils.ts:10]: `bot.config.selfUrl` 是你在 bot.ts 里写的 `Schema` 中定义的 `selfUrl`，然后你可以从 `ctx.options.selfUrl` 拿到全局的 `selfUrl`。优先使用 bot 自己的 `selfUrl`，如果不存在再使用全局的

File: adapters/feishu/src/index.ts (hunk 1 above)
Comment (Anillc) [reply adapters/feishu/src/index.ts:10]: 因为用户可能会希望处理一些别的事件，你得把事件的详细内容也加到 session 里。可以参考别的 adapter 也是这样做的，比如 onebot 中加入的是 Payload

File: adapters/feishu/src/bot.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Anillc) [reply adapters/feishu/src/bot.ts]: `adaptSession` 看上去可以去掉 `async` 了，另外 `dispatchSession` 那边的返回值也没有去掉 `Promise`

File: adapters/feishu/src/index.ts (hunk 1 above)
Comment (MaikoTan) [reply adapters/feishu/src/index.ts:10]: 如此的話，何不在 `interface Event {}` 中加入自定義事件，如 `feishu/interactive` 為了活動卡片。

File: adapters/feishu/src/utils.ts (hunk 2 above)
Comment (MaikoTan) [reply adapters/feishu/src/utils.ts:10]: `bot.ctx.options.selfUrl` 的類型顯示為 `(property) Context.Config.Network.selfUrl?: string`，若用戶未配置此處同上文的 `bot.config.selfUrl`，那麼 `bot.config.selfUrl ?? bot.ctx.options.selfUrl` 是否可以拿到可用的值？

Comment (Anillc) [reply adapters/feishu/src/utils.ts]: 不一定，因为全局的 `selfUrl` 也可能没有配置，所以我认为需要加个判断，如果没有可用的 `selfUrl` 就发一个警告

File: adapters/feishu/src/index.ts (hunk 1 above)
Comment (Anillc) [reply adapters/feishu/src/index.ts]: 主要是考虑到别的 adapter 也是这样设计的。而且 session 本来就会进入事件，在 `dispatch` 的时候事件就会发出去，所以在加入 `& Event` 的同时，也应该在 `interface Event {}` 中加入事件

https://github.com/satorijs/satori/blob/a3ba417bdc99f5df9cfda6a9b1b0f7a1d79b7095/packages/core/src/bot.ts#L94-L106

File: adapters/feishu/src/utils.ts (hunk 2 above)
Comment (MaikoTan) [reply adapters/feishu/src/utils.ts]: Technically, if you want to retrive messages from feishu, a public IP address is required. So there may always be a `selfUrl` in the global settings.
Donc, I think it is possible to check the `selfUrl` options at adapter initialisation, if it was not set, then report a warning. But what do you think that how the asset URL would be generated in that case?

Comment (Anillc) [reply adapters/feishu/src/utils.ts]: 听上去没问题

File: adapters/feishu/src/bot.ts [hunk 6]
```diff
@@ -236,6 +236,44 @@ export namespace FeishuBot {
...
+    export type Contents<T extends Content['msg_type']> = Content & (
+      T extends 'text' ? Text :
+      T extends 'image' ? Image :
+      T extends 'audio' ? Audio :
+      T extends 'media' ? Media :
+      T extends 'file' ? File :
+      Content
+    )
+    export type FileContents = Content & (Image | Audio | Media | File)
+    export function isWhatContents<T extends Content['msg_type']>(
```
Comment (MaikoTan) [inline adapters/feishu/src/bot.ts:270]: naming suggest: `isContentType`

File: adapters/feishu/src/modulator.ts [hunk 7]
```diff
@@ -2,7 +2,14 @@ import { Modulator, segment } from '@satorijs/core'
...
 
 export class FeishuModulator extends Modulator<FeishuBot> {
-  private buffer: any
+  private mode: 'default' | 'figure' = 'default'
+  private content = ''
+  private addition = {
+    file: null as {
+      id: string
+      type: FeishuBot.Message.FileContents['msg_type']
+    } | null,
```
Comment (MaikoTan) [inline adapters/feishu/src/modulator.ts:11]: Personally I don't like `null`, but prefer `undefined` to refer to a nil value.

Comment (shigma) [reply adapters/feishu/src/modulator.ts:11]: Actually you can omit this `| null` because we do not use SNC.

Comment (MaikoTan) [reply adapters/feishu/src/modulator.ts:11]: True, but I am also refering the `file: null as {...}` part, which can be just `undefined` in this case.

File: adapters/feishu/src/bot.ts [hunk 8]
```diff
@@ -236,6 +236,44 @@ export namespace FeishuBot {
...
+      type: T, data: Contents<Content['msg_type']>
+    // @ts-ignore
+    ): data is Contents<T> {
+      return data.msg_type === type
+    }
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply adapters/feishu/src/bot.ts]: Removed as duplicated. Sorry @NWYLZW

[the remark this answers is not in the document]
Comment (NWYLZW) [reply adapters/feishu/src/bot.ts]: 才收到消息，哈哈哈（

File: adapters/feishu/src/message.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        },
+      }
+    }
+  }
+
+  async visit(element: segment) {
+    const { type, attrs, children } = element
+
+    // render rich text post
+    if (this.mode === 'figure') {
```
Comment (MaikoTan) [inline adapters/feishu/src/message.ts:127]: Should I reset `this.mode` to `'default'` somewhere to make sure the next elements render in a right way?

File: adapters/feishu/src/message.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+  file: MessageContent.MediaContents
+  type: MessageType
+}
+
+export class FeishuMessenger extends Messenger<FeishuBot> {
+  private mode: 'default' | 'figure' = 'default'
+  private quote: string | undefined
+  private content = ''
+  private addition: Addition
+  private richText: MessageContent.RichText[string]
```
Comment (MaikoTan) [inline adapters/feishu/src/message.ts:20]: Is there a better typing, as we know that Feishu's rich text elements could contain multiple languages.

File: adapters/feishu/src/message.ts [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+  async post(data?: any) {
+    try {
+      let resp: BaseResponse & { data: Message }
```
Comment (MaikoTan) [inline adapters/feishu/src/message.ts:24]: Also here, the typing is strange.

File: adapters/feishu/src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+export type Sender =
+  | {
+      sender_id: Feishu.UserIds;
+      sender_type?: string;
+      tenant_key: string;
+    }
+  | (Feishu.UserIdentifiers & { sender_type?: string; tenant_key: string });
+
+export function adaptSender(sender: Sender, session: Session): Session {
```
Comment (MaikoTan) [inline adapters/feishu/src/utils.ts:16]: There's only `open_id` you received in a message, should I get their information by requesting other APIs to make sure those being set in the `session` object?

File: adapters/feishu/src/message.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    // render rich text post
+    if (this.mode === 'figure') {
+      if (!this.richText) {
+        this.richText = { title: '', content: [] }
+      }
+      switch (type) {
+        case 'text':
+          this.richText.content.push([{ tag: 'text', text: attrs.content }])
+          break
+        case 'link':
```
Comment (MaikoTan) [inline adapters/feishu/src/message.ts:135]: `link` can be only sent as rich text elements, or should we just extract the URL and then send it as plain text?

**Review by MaikoTan — COMMENTED** [review]
@NWYLZW Can you review once again?

File: adapters/feishu/src/message.ts (hunk 10 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts:20]: 对于多语言来说，我觉得应该先让 satori 来做（比如定义一个规范之类的）
虽然飞书能自然而然的支持多语言，但是如果先支持这个会导致后面的改造成本会比较高

Comment (NWYLZW) [reply adapters/feishu/src/message.ts:20]: 另外如何让他和目前的 i18n 兼容呢，用户应该怎么去设置自己的语言呢，优先级又该怎么确定呢
所以我觉得还是以 i18n 目前的逻辑为准，不在单个适配器上做过多的工作

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:20]: Yeah, so I currently just using `zh_cn` as the default locale of a rich text post.

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:20]: Also, according to kawaii @shigma , the `embed elements` (rich text in feishu, though) would have a nouveau implementation in the future.
So my opinion is not going to make it complicated, so we can adjust it in the next days.

Comment (NWYLZW) [reply adapters/feishu/src/message.ts:20]: 对于富文本消息我倒是没有什么疑惑的点
最主要的还是 satori 是否需要去做 i18n 的适配
是否有更好的方案设计

File: adapters/feishu/src/message.ts (hunk 13 above)
Comment (MaikoTan) [reply adapters/feishu/src/message.ts]: Looks like that `a` is the correct one to refer to a hyper link, so I changed at the bottom.

**Comment by MaikoTan** [comment]
OK, Let's give it a try!

File: adapters/feishu/src/message.ts (hunk 9 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts:127]: 我觉得你现在写的有点不一样，我好像在 L175 和 L178 就处理了这部分逻辑

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:127]: You refer to this lines? Looks like they do cover the problem I mentioned.

https://github.com/satorijs/satori/blob/d3e2e4552297711bbf30575c55e11de949b82146/adapters/feishu/src/message.ts#L176-L178

Comment (NWYLZW) [reply adapters/feishu/src/message.ts:127]: 我不太确定 figure 的行为了，这个地方应该让 @shigma 确认一下

一、figure 在不支持的平台上应该怎么去定义行为
二、figure 如果要支持，那么行为应该是什么样子的
三、figure 这个名字是不是要替换为 forward

我感觉你现在的实现和 figure 的语义不太一样（我不确定）

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:127]: Well, I asked @shigma -chan last night, she told me that the figure should be implemented to send messages in one message as much as possible. So I use rich text post to achieve that.

File: adapters/feishu/src/message.ts (hunk 11 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts:24]: 你看起来已经改了，要不要 resolve 掉

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:24]: Oui !

File: adapters/feishu/src/message.ts (hunk 9 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts:127]: 意思就是说，figure 应该尽可能提供一个完整的消息

那么 message 类型消息的行为呢？是否要定义为，能一个消息发出来就发出来，发不出来就分片发送

不然实际上对于飞书来说，message 类型和 figure 类型就是完全一致的行为了（或者说这就是预期的）

如果要一样的行为，这部分逻辑是否可以统一

File: adapters/feishu/src/utils.ts (hunk 12 above)
Comment (NWYLZW) [reply adapters/feishu/src/utils.ts:16]: 你是说的 at 行为的时候用户名吗，这个是一个问题点
我觉得是需要请求额外的 api 吧，这个地方是不是不是一个很急的需求（我是这么想的，他的优先级并不高）
关于这个我有点问题
一、一个消息中可能出现多个重复 at 吗
二、用户名的信息应该从 satori/kosihi 拿，还是 api 优先级最高

File: adapters/feishu/src/message.ts (hunk 13 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts]: 如果 link 只能通过 rich 的方式发送，那么所有的消息应该都用 rich 去保证解析正确吗

我觉得成本有点高，就如我在上面提到的一点，是不是要区分 figure 的行为（还是把所有的消息作为富文本消息）

（或者当消息中出现了只有富文本支持的消息的时候再触发富文本模式，figure 的语义看起来并不符合场景）

File: adapters/feishu/src/message.ts (hunk 9 above)
Comment (MaikoTan) [reply adapters/feishu/src/message.ts:127]: Well, no.
A `message` element can be split.

<details><summary>Chat Log</summary>



</details>

File: adapters/feishu/src/utils.ts (hunk 12 above)
Comment (MaikoTan) [reply adapters/feishu/src/utils.ts:16]: No, the function here is aim to extract the user id of the sender in a message received event.
Maybe there should be an API to update user info later, but not in this PR in my opinion.

File: adapters/feishu/src/message.ts (hunk 13 above)
Comment (MaikoTan) [reply adapters/feishu/src/message.ts]: Yep, I thought rich text is also heavy here. So I now follow the onebot way to just send a URL with parens.

File: adapters/feishu/src/message.ts (hunk 9 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts:127]: 其实这个我知道，我应该是把 figure 和 message 弄混了

我先捋捋看

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:127]: image

File: adapters/feishu/src/message.ts (hunk 13 above)
Comment (NWYLZW) [reply adapters/feishu/src/message.ts]: 那这个地方的行为对我来说我没什么别的看法了，我觉得是正确的行为

File: adapters/feishu/src/utils.ts (hunk 12 above)
Comment (NWYLZW) [reply adapters/feishu/src/utils.ts:16]: 那这个地方的问题，我没有搞清楚，我不清楚这个地方有什么场景会有问题

Comment (MaikoTan) [reply adapters/feishu/src/utils.ts:16]: You see there are many properties that plugins might need for.
Par ex. many plugins need `username` or `nickname`, some manager plugins need `roles`, also assignment need `isBot` I think?
So is there any requirement to have them set up?

File: adapters/feishu/src/message.ts [hunk 14]
```diff
@@ -150,37 +150,50 @@ export class FeishuMessenger extends Messenger<FeishuBot> {
...
+        this.mode = 'figure'
+        await this.render(children, true)
+        this.mode = 'default'
+        break
+      case 'message':
+        await this.flush()
+        await this.render(children, true)
+        break
+      default:
+        await this.render(children)
```
Comment (shigma) [inline adapters/feishu/src/message.ts:196]: 学到了，下次我也用这种方法凑行数 🤣

Comment (MaikoTan) [reply adapters/feishu/src/message.ts:196]: lmao 🤣

**Pull request merged**